src/java/org/apache/cassandra/utils/memory/BufferPool.java

```diff
@@ -777,41 +817,39 @@ private void put(ByteBuffer buffer, Chunk chunk)
                 return;
             }
 
-            // ask the free method to take exclusive ownership of the act of recycling if chunk is owned by ourselves
-            long free = chunk.free(buffer, owner == this && recycleWhenFree);
-            // free:
-            // *     0L: current pool must be the owner. we can fully recyle the chunk.
-            // *    -1L:
-            //          * for normal chunk:
-            //              a) if it has owner, do nothing.
-            //              b) if it not owner, try to recyle it either fully or partially if not already recyled.
-            //          * for tiny chunk:
-            //              a) if it has owner, do nothing.
-            //              b) if it has not owner, recycle the tiny chunk back to parent chunk
-            // * others:
-            //          * for normal chunk:  partial recycle the chunk if it can be partially recycled but not yet recycled.
-            //          * for tiny chunk: do nothing.
-            if (free == 0L)
+            long free = chunk.free(buffer);
+
+            if (free == -1L && owner == this && owningThread == Thread.currentThread() && recycleWhenFree)
```

Comment:
Why isn't `owner == this` not sufficient, and if it isn't, is there a performance overhead by calling `Thread.currentThread()` every time when returning buffers to the pool?

Reply:
The check happens only on recycling the whole chunk, not on returning every single buffer. AFAIK the check itself is also relatively lightweight.